session/session.go

```diff
@@ -1045,6 +1045,101 @@ func (s *session) Execute(ctx context.Context, sql string) (recordSets []sqlexec
 	return
 }
 
+type sessionVarsRevertInfo struct {
+	varName string
+	datum   types.Datum
+}
+
+func getHintDatumFromBool(flag bool) types.Datum {
+	if flag {
+		return types.NewStringDatum("on")
+	}
+	return types.NewStringDatum("off")
+}
+
+func (s *session) setSessionVarsByHints(stmtNode ast.StmtNode) (revertInfos []sessionVarsRevertInfo, warns []error) {
+	var hints []*ast.TableOptimizerHint
+	switch x := stmtNode.(type) {
+	case *ast.SelectStmt:
+		hints = x.TableHints
+	case *ast.UpdateStmt:
+		hints = x.TableHints
+	case *ast.DeleteStmt:
+		hints = x.TableHints
+	// TODO: support hint for InsertStmt
+	default:
+		return
+	}
+	var memoryQuotaHintList, noIndexMergeHintList, useToJAHintList []*ast.TableOptimizerHint
+	for _, hint := range hints {
+		switch hint.HintName.L {
+		case "memory_quota":
+			memoryQuotaHintList = append(memoryQuotaHintList, hint)
+		case "no_index_merge":
+			noIndexMergeHintList = append(noIndexMergeHintList, hint)
+		case "use_toja":
+			useToJAHintList = append(useToJAHintList, hint)
+		}
+	}
+	// Handle MEMORY_QUOTA
+	if len(memoryQuotaHintList) > 1 {
+		warn := errors.New("There are multiple MEMORY_QUOTA hints, only the last one will take effect")
+		warns = append(warns, warn)
+	}
+	if len(memoryQuotaHintList) != 0 {
+		hint := memoryQuotaHintList[len(memoryQuotaHintList)-1]
+		// Executor use MemoryQuota < 0 to indicate no memory limit, here use it to handle hint syntax error.
+		if hint.MemoryQuota < 0 {
+			warn := errors.New("There are some syntax error in MEMORY_QUOTA hint, correct usage: MEMORY_QUOTA(10 M) or MEMORY_QUOTA(10 G)")
+			warns = append(warns, warn)
+		} else {
+			revertInfos = append(revertInfos, sessionVarsRevertInfo{
+				varName: variable.TIDBMemQuotaQuery,
+				datum:   types.NewIntDatum(s.sessionVars.MemQuotaQuery),
+			})
+			err := variable.SetSessionSystemVar(s.sessionVars, variable.TIDBMemQuotaQuery, types.NewIntDatum(int64(hint.MemoryQuota)))
```

Comment:
Since all hints should be available inside a SQL (SQL-level), it seems not appropriate to put them in the session variable struct.
How about using `StmtCtx.SetValue` to store these SQL-level variables.

Reply:
And I thought it should work with query block...